contracts/relayer/BaseRelayer.sol

```diff
@@ -0,0 +1,61 @@
+pragma solidity ^0.4.24;
+
+import "../lib/sig/ECDSA.sol";
+import "../apps/AragonApp.sol";
+import "../common/DepositableStorage.sol";
+
+
+contract BaseRelayer is AragonApp, DepositableStorage {
+    using ECDSA for bytes32;
+
+    bytes32 public constant OFF_CHAIN_RELAYER_SERVICE_ROLE = keccak256("OFF_CHAIN_RELAYER_SERVICE_ROLE");
+
+    uint256 private constant EXTERNAL_TX_COST = 21000;
+
+    string private constant ERROR_GAS_REFUND_FAIL = "RELAYER_GAS_REFUND_FAIL";
+    string private constant ERROR_NONCE_ALREADY_USED = "RELAYER_NONCE_ALREADY_USED";
+    string private constant ERROR_INVALID_SENDER_SIGNATURE = "RELAYER_INVALID_SENDER_SIGNATURE";
+
+    event FundsReceived(address indexed sender, uint256 amount);
+    event TransactionRelayed(address indexed from, address indexed to, uint256 nonce, bytes calldata);
+
+    modifier refundGas() {
+        uint256 startGas = gasleft();
+        _;
+        uint256 refund = EXTERNAL_TX_COST + startGas - gasleft();
```

Comment:
Two comments about `EXTERNAL_TX_COST`:

- If the sender is a contract, they would be able to call into this contract multiple times in one transaction, so we would be refunding more than necessary.
- Assuming that the call is performed by an EOA, the cost will always be higher than 21k gas as the contract needs to read the calldata and find the function that is being called before the modifier sets the `startGas`.

Reply:
Sure, I just wanted to provide the closest  implementation to have an idea of its costs. We will need to do the maths and write some tests as well.

Reply:
I'm actually working on another approach cause this one has a ton of edge cases and it actually costs a lot of gas to calculate how much gas we need to refund